paddle/fluid/memory/allocation/legacy_allocator.cc

```diff
@@ -168,9 +170,9 @@ BuddyAllocator *GetGPUBuddyAllocator(int gpu_id) {
                << ". Current 'FLAGS_reallocate_gpu_memory_in_mb' value is "
                << FLAGS_reallocate_gpu_memory_in_mb << "\n\n";
     }
+    platform::SetDeviceId(gpu_id);
   });
 
-  platform::SetDeviceId(gpu_id);
```

Comment:
Don't need to invoke `platform::SetDeviceId(gpu_id);` every time when calling `GetGPUBuddyAllocator `.